### PR TITLE
Implementing basic optimizer and test bench

### PR DESCRIPTION
**Pull request by ghost**
## Context:
Optimizer project for porting coil optimizer modules for different shimming goals.

## Goals:
For this PR, 

- [x] Slice by slice sequencer
- [x] Basic optimizer using least squares
- [x] Generate coil profiles for arbitrary loops
- [x] Tests for all functions

## Future work:
In future PRs, amore complex sequencer that deals with realtime modifications will follow that. 

Refers to: #20 #30

### CODE REVIEW

**Comment by po09i** [comment]
oops, I think I messed up some git rebase stuff, not sure what went wrong or how to fix it.

**Comment by po09i** [comment]
Ok pretty sure I fixed it

**Review by po09i — CHANGES_REQUESTED** [review]
Seems like you included the log in you commit, we probably want to remove that.

**Comment by lcbMGH** [comment]
I need help with Travis (haven't used it before)-- once that is good, this PR is ready for review

**Comment by jcohenadad** [comment]
> I need help with Travis (haven't used it before)-- once that is good, this PR is ready for review

looks like the [unit test `test_zslice()` is not passing](https://travis-ci.com/github/shimming-toolbox/shimming-toolbox-py/builds/185355460#L470). Are you able to reproduce this error when running unit test locally?

**Comment by jidicula** [comment]
> I need help with Travis (haven't used it before)-- once that is good, this PR is ready for review

If you dig into the [build log](https://travis-ci.com/github/shimming-toolbox/shimming-toolbox-py/builds/185355460), it looks like `test_zslice` failed since the commits 5ff0d78, 3527f2a, da78edf, and ab51bd8 that @po09i added.

**Comment by jcohenadad** [comment]
the error is likely related to the fact that z is(100,100,3) and i_slice is a scalar.

**Comment by lcbMGH** [comment]
Oh, I'm at the level of not getting to run unit tests locally. Currently, it's crashing through missing the examples module -- working on that. I appreciate you running through the error yourselves!

File: shimmingtoolbox/coils/biot_savart.py
Comment (jidicula) [inline shimmingtoolbox/coils/biot_savart.py]: Since these are constants, it's good practice to make them all-caps

**Review by jidicula — DISMISSED** [review]
Really small styling change requested.

File: shimmingtoolbox/coils/biot_savart.py
Comment (po09i) [reply shimmingtoolbox/coils/biot_savart.py]: Fixed in commit 9ee12e3

**Review by po09i — DISMISSED** [review]
There is currently no test for the bio_savart coil generation. A simple test should be implemented to test it.

The coil profile binaries for bio-savart and spherical_harmonics added in 9fc6608 should not be in this repo. They are used for data testing and should go in the [data-testing repo](https://github.com/shimming-toolbox/data-testing). Refers to #82. Moving the coil profiles to the testing repo and removing them from this PR seems like the proper way to go. Before I move the coil profiles, since I did not generate them I am not familiar with them or the script used to do so, @lcbMGH could you fill in the details in issue #82. The spherical_harmonics coil_profile is also really quick to generate so it should not be included in the data-testing in my opinion. DONE

File: shimmingtoolbox/coils/biot_savart.py
Comment (jcohenadad) [inline shimmingtoolbox/coils/biot_savart.py:12]: for the unit test, we could have a unit test that checks a vector (hard-coded in the unit test), instead of needing to add a binary file for testing.

also, the test should be quick (ie: small number of segments, small matrix)

**Comment by lcbMGH** [comment]
We should rebase and remove the coil_examples

**Comment by po09i** [comment]
> We should rebase and remove the coil_examples

Done

**Comment by lcbMGH** [comment]
I screwed up the merge, and brought the coil example profiles back -- I'll work on fixing this, otherwise it's ready for checks

**Comment by po09i** [comment]
> I screwed up the merge, and brought the coil example profiles back -- I'll work on fixing this, otherwise it's ready for checks

We need to rebase again, not a problem I can do it since it's really similar to the last rebase. The problem is you can't `git pull` after a rebase you need to do something different since the history was changed. I'll post the proper way to do so after the rebase.

**Comment by lcbMGH** [comment]
Branch is now ready for final review -- make sure to run `git fetch` and `git reset --hard origin/ad-lcb/20-optimizer` to avoid merge issues.

**Review by po09i — COMMENTED** [review]
I changed everything I would've commented on but since I was one of the main contributors I'll wait for somebody else to approve the PR.

**Review by jcohenadad — APPROVED** [review]
LGTM but i haven't tested.
wonderful work 🏅

**Comment by ghost** [comment]
I'll wait for one more approval, but I'm going to merge this before the end of the month if there's no other concerns!

**Comment by po09i** [comment]
I had to fix a couple of issues so that we can merge with master that were caused by errors with rebasing from Lincoln and me during the PR. Should be good to go now.

I had to rebase so make sure to `git fetch and git reset --hard origin/ad-lcb/20-optimizer`

**Comment by jcohenadad** [comment]
please "squash and merge" as now agreed in #14

**Review by po09i — APPROVED** [review]

**Pull request merged**
